### PR TITLE
Zombie Mode+

### PR DESCRIPTION
**Pull request by Super3222**
### Intent of your Pull Request

- If the population (people on station will be affected - ghost roles don't count unless they're on the station; late joiners will be added to the population list) is dropped to 10%, **the shuttle is called, and cannot be recalled.** I am willing to drop to 5% or increase this to 15% depending on in game statistics, and how future development goes.
- Zombies now have a dead on 2 minutes before they turn. It use to be randomized between 1 minute - 2.30 minutes. If you need to compare this time to something do it to blob, which is 3 minutes before burst.
- Zombies now speak zombie. They no longer understand human.
- **Balance notes for everything else is in one of my last posts**

I'm thinking of giving roundstart zombies a "sense human" ability, that allows them to pick up the scent of the nearest NON-INFECTED human. So you don't get slow ass rounds.

#### Changelog

:cl:
experiment: WOOHOO! ZOMBIE MODE BALANCE CHANGES! ASK YOUR COUNCIL TODAY TO VOTE ZOMBIE MODE INTO ROTATION!
experiment: During zombie mode if the human population is dropped to 10% the shuttle will be automatically called. Nobody will be able to recall it
experiment: Roundstart zombies have a specific amount of time before they turn. Beforehand it was randomized.
experiment: Zombies no longer understand human.
experiment: If a zombie is set on fire after it dies it will burn into ashes and cease to exist. Everything they carried will drop on the ground, and be set on fire if possible.
experiment: Zombie claw damage dropped from 20 -> 15. It will most likely be picked back up in the future.
experiment: Zombies can now heal from attacking human corpses. There is a limit however, the corpse must have above -200 health.
experiment: Zombies no longer heal passively.
experiment: Zombies will be able to heal from knocked over targets (ie pulling out their intestines and devouring them).
experiment: Zombies can now open firelocks.
experiment: Nerfed the damage the zombie organ does, but makes it stronger the longer you have it in you.
experiment: The chance of a zombie claw infecting someone now runs through how much armor the target has on, and where the zombie is targeting.
/:cl:

------

So let's start theorizing what might happen from this change.

population size | 10% of population (roundend)
7 (base)-10 | 1
20 | 2
30 | 3
40 | 4
50 | 5
60 | 6

As you can see these are pretty low numbers for a LARGE amount of people that will either be A - hunting your ass or B - waiting for the round to end, which is why a change like this makes sense to happen.

Beforehand without this change, the mode relied on two things to happen.
1 - It required EVERY LAST human to either be zombified or DIE.
2 - It also required the shuttle to reach centcom, which a lot of people didn't manage to accomplish when Slinky was running the round.

Since neither of those two things happened last time, Adam was forced to end the round because it was becoming unbearable.

Now, the problem with roundstart zombies is that they're more prone to run around an grief before turning into a zombie compared to blobs. As a blob, if you go around and grief you'll get some kills, but you'll burst in a bad location and most likely get fucked. As a zombie, if you go around and grief you'll most likely get enough body counts to bring back an army of the dead before you turn.

Which makes my decision to make reanimation time for zombies take 2 minutes tops instead of 3 like blob more logical. Making it one minute would be ridiculous because that sort of time limit puts too much pressure on them.

### CODE REVIEW

**Comment by Time-Green** [comment]
Could you perhaps split slashing and biting? 
Slashing being slashing and biting being on a timer, so it's not instant

**Comment by AlpacaGoesCrazy** [comment]
Maybe disallow zombies to listen to comms? Moaning for brains on comms are fun but listening to chatting humans for meta is kinda wrong.
Also zombies should smash things more effectively (firelocks, maybe regular walls etc.), but they should not be able to haul stuff around (real zombies would be too dumb for that)

**Comment by Super3222** [comment]
> Could you perhaps split slashing and biting?
> Slashing being slashing and biting being on a timer, so it's not instant

I don't want to. At least, not right now

However you could either port /vg/stations bite hud, make it an action button, or make it trigger on the AltClick (check click.dm for that).

> Maybe disallow zombies to listen to comms? Moaning for brains on comms are fun but listening to chatting humans for meta is kinda wrong.

I could do that honestly. Just give me a second. 

> Also zombies should smash things more effectively (firelocks, maybe regular walls etc.), but they should not be able to haul stuff around (real zombies would be too dumb for that)

I agree with firelocks, but not regular walls.

**Comment by AlpacaGoesCrazy** [comment]
Anyway zombie mechanics should encourge 'dumb zombie horde rush humans crush everything' kinda gamestyle instead of 'lurking solo pro mlg zombie 360 noscope biting humans and being robust'

**Comment by Super3222** [comment]
> Anyway zombie mechanics should encourge 'dumb zombie horde rush humans crush everything' kinda gamestyle instead of 'lurking solo pro mlg zombie 360 noscope biting humans and being robust'

Yeah, I'll keep that in mind.

**Comment by ShadowDeath6** [comment]
Could be interesting to look at https://github.com/yogstation13/yogstation/pull/2265 , I wouldn't take away zombie healing but it would be a good idea to cut the toxins damage caused by the organ, making it half as likely to damage would be good

**Comment by oisin100** [comment]
Make them actually not be impervious to plasma fire.

Make them turn to ashes if burnt to death(Although not lasered), I want to purge the station of zombies with plasma flooding.

**Comment by Super3222** [comment]
@oisin100 I'll do that after Xeno Mode and Vampire Mode get reviewed

**Comment by Super3222** [comment]
and I've done it, but you'll have to wait until I re open this PR.

**Comment by ghost** [comment]
Super, a few issues with zombies-

One hit means you are infected. While this makes enough sense for a guy in no clothes, it makes very little sense for a guy in a hardsuit or significant armor-and in fact makes it useless against zombies since one hit = you're fucked.

Zombies still heal over time and one scratch means the infected person dies -very quickly-. It's effectively impossible to win against zombies.

Conversely, zombies cannot get past walls and firelocks and such-meaning the crew can win simply by making a large box around the bar and staying in it, and nothing can ever get in again.

**Comment by Super3222** [comment]
- Lowered brute damage of zombie claw from 20 to 15.
- Lowered the chance of being damaged by a zombie organ to 5%, and the amount of toxin damage you take depends on how long you've had the organ (capped at 30). Remember that this is checked literally EVERY TICK. We want the person to get sick from the zombie organ and die gradually. Not IMMEDIATELY.
- Removed zombies healing passively
- Refactored handle_tear() - called when someone rips apart of their jumpsuit.
- Zombies now heal from attacking dead bodies, however, that dead body must be above -200 health. Every time this happens, the zombie will regain 5-10 health and the target will be dealt 30 brute damage.
- Zombies can also feast on humans that are lying down, ripping out their intestines, and devouring them for health. Keep in mind that zombies will still deal 30 brute damage to them.
- Zombies can now open firelocks.
- Zombies infection system has been completely refactored it works like this:

* Zombie attacks someone.
* If the zombie is aiming at the head/eyes/mouth, whatever the person has on is thrown off.
* In the case that it's a hardsuit helmet, a 45% chance is rolled to see if the helmet disengages.
* In case that our helmet/hat has armor values, we roll for a chance to block the claw attack depending on those values. If it fails, then we get infected.
* If the zombie is aiming anywhere else, we at first tear off whatever armor they're wearing and roll a chance to infect the person depending on how powerful their armor is (this is because they could easily pick the armor back up, and put it on)
* If they aren't wearing any armor their uniform takes 80 damage, and there's only a 20% chance to become infected. Another swipe will shred it to pieces, and leave a 100% chance to infect.

With this system, putting on armor has a better chance at protecting you from being hit. As long as the armor is good. So far it only checks for brute damage, but if anyone wants to change that in the future be my guest. Maybe with armor that contains bio protection, there should be a 100% chance to block an infection.

in short:
Zombie claws will STILL hurt people, but the chance for infection is rolled depending on how strong the targets armor is. In the case that the person is lying down even MORE damage is dealt to the target.

**Comment by Super3222** [comment]
Also, remind me to give them stun and stamina immunity.

**Comment by Super3222** [comment]
FUCK I didn't mean to close that

**Comment by ShadowDeath6** [comment]
zombies should not be able to rip apart hardsuits and things like riot armor, I don't care how strong they are

**Comment by Super3222** [comment]
They don't rip apart hardsuits.

**Comment by Super3222** [comment]
Alright, the plan was to make it so that clothes would have a chance of preventing you from getting infected. It takes 2 attacks from a zombie claw to successfully tear your uniform to shreds. If you had armor / something on covering you, it will force strip that off first instead.

EDIT: Also it doesn't destroy hardsuits @ShadowDeath6 It force-strips them off

**Comment by ghost** [comment]
not riot armor either

being able to infect you instantly through 5 layers of protection is dumb tbh

i'd prefer if they could only infect critted people

File: code/modules/clothing/clothing.dm [hunk 1]
```diff
@@ -378,32 +390,31 @@ BLIND     // can't see anything
...
-			//no more cloth left on the item, so nix it
-			user.visible_message("[src] falls away to tatters, stripped to its barest seams.")
-			qdel(src)
-			if (ishuman(user))
-				var/mob/living/carbon/human/H = user
-				H.update_inv_w_uniform()
+
+		var/new_ripcount = ripcount - 1
+		if(ripcount)
+			handle_tear(user, ripcount = new_ripcount)
```
Comment (X-TheDark) [inline code/modules/clothing/clothing.dm]: Why is it necessary to call this proc recursively?

File: code/modules/zombie/items.dm [hunk 2]
```diff
@@ -31,34 +30,108 @@
...
 /obj/item/zombie_hand/proc/check_feast(mob/living/target, mob/living/user)
-	if(target.stat == DEAD)
-		var/hp_gained = target.maxHealth
-		target.gib()
-		user.adjustBruteLoss(-hp_gained)
-		user.adjustToxLoss(-hp_gained)
-		user.adjustFireLoss(-hp_gained)
-		user.adjustCloneLoss(-hp_gained)
-		user.adjustBrainLoss(-hp_gained) // Zom Bee gibbers "BRAAAAISNSs!1!"
+	if(target.stat == DEAD && target.health > -200 || target.resting)
```
Comment (X-TheDark) [inline code/modules/zombie/items.dm]: `if( (target.stat == DEAD && target.health > -200) || target.resting)`
or
`if( target.stat == DEAD && (target.health > -200 || target.resting) )`

BYOND has really shitty way or resolving the priorities of `&&` and `||` operators, so if you mix them, please enclose them in brackets. This is the kind of thing that caused exponential stock price increase.

File: code/modules/clothing/clothing.dm (hunk 1 above)
Comment (Super3222) [reply code/modules/clothing/clothing.dm]: `ripcount` is a variable that I'm using to measure how many times we want to rip the clothing.

Comment (Cruix) [reply code/modules/clothing/clothing.dm]: But why is it called recursively instead of iteratively?

**Comment by Super3222** [comment]
Zombie Paw is incredibly broken, and results in crashing. Probably due to an infinite loop I created. I'm trying to solve the problem right now.

**Comment by Super3222** [comment]
Zombie Claw is incredibly broken, and results in crashing. Probably due to an infinite loop I created. I'm trying to solve the problem right now.

**Comment by Super3222** [comment]
This is ready for testing or what have you. The premise is zombies need to claw off clothing to infect people. Zombie claws are accommodated to force strip stuff off, but will have a smaller success rate if the clothing has armor (ie. riot armor is harder to claw off).

**Comment by Time-Green** [comment]
Works good. Wasn't able to test most gamemode features for obvious reasons
Also the o2 icon flashes for some reason wich is kinda annoying but I guess it's fine considering lavaland has it aswell, and zombies dont breathe so technicly they suffocate
